uk_geo_utils/base_updater.py

```diff
@@ -0,0 +1,377 @@
+                    OR
+                    confrelid = '{self.table_name}'::regclass
+                )      
+        """
+        self.cursor.execute(fkey_sql)
+        results = self.cursor.fetchall()
+
+        self.stdout.write(
+            f"Found {len(results)} foreign key constraints, where {self.table_name} is the referencing or referenced table"
+        )
+
+        fk_constraints = []
+
+        for row in results:
+            constraint_name = row[0]
+            # referenced_table_name = row[1]
+            constraintdef = row[2]
+            referencing_table = row[3]
+
+            fk_constraints.append(
+                {
+                    "constraint_name": constraint_name,
+                    "create_statement": f"ALTER TABLE {referencing_table} ADD CONSTRAINT {constraint_name} {constraintdef} NOT VALID",
+                    "delete_statement": f"ALTER TABLE {referencing_table} DROP CONSTRAINT IF EXISTS {constraint_name}",
+                    "validate_statement": f"ALTER TABLE {referencing_table} VALIDATE CONSTRAINT {constraint_name}",
+                }
+            )
+
+        return fk_constraints
+
+    def drop_foreign_keys(self):
+        self.stdout.write("Dropping foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            self.stdout.write(f"Executing: {constraint['delete_statement']}")
+            self.cursor.execute(constraint["delete_statement"])
+
+    def add_foreign_keys(self):
+        self.stdout.write("Creating foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            try:
+                self.stdout.write(
+                    f"Executing: {constraint['create_statement']}"
+                )
+                self.cursor.execute(constraint["create_statement"])
+            except (
+                ProgrammingError
+            ):  # better if psycopg2.errors.DuplicateObject:
+                self.stdout.write(
+                    f"{constraint['constraint_name']} already exists."
+                )
+
+    def validate_foreign_keys(self):
+        self.stdout.write("Validating foreign keys...")
+        for constraint in self.foreign_key_constraints:
+            self.stdout.write(f"Executing: {constraint['validate_statement']}")
+            self.cursor.execute(constraint["validate_statement"])
+
+    def create_temp_table(self):
+        self.stdout.write(
+            f"Creating temp table called {self.temp_table_name}..."
+        )
+        self.cursor.execute(f"DROP TABLE IF EXISTS {self.temp_table_name};")
+        self.cursor.execute(
+            f"CREATE TABLE {self.temp_table_name} AS SELECT * FROM {self.table_name} LIMIT 0;"
+        )
+
+    def index_exists(self, index_name):
+        self.cursor.execute(f"""
+            SELECT 1 
+            FROM pg_indexes 
+            WHERE indexname = '{index_name}'
+        """)
+
+        return bool(self.cursor.fetchone())
+
+    def rename_old_table(self):
```

Comment:
I don't think I totally get why we're doing the rename to `_old` thing on tables/indexes. What does it buy us over deleting and renaming inside a transaction?

Reply:
Discussed in comment about foreign key validation

Reply:
No longer using the old tables. See fixup: 978bf3fc23b35b